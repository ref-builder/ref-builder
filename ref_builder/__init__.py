"""Virtool CLI for managing references.

TODO: Switch back to Rich.
TODO: Allow exclusion checks at the Repo level.
TODO: Don't use the term "blacklist". Use the word "exclusions" only.
"""
