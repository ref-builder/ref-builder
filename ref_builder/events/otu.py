from pydantic import UUID4

from ref_builder.events.base import EventData, Event, OTUQuery
from ref_builder.models import Molecule
from ref_builder.plan import MonopartitePlan, MultipartitePlan


class CreateOTUData(EventData):
    """The data for a :class:`CreateOTU` event."""

    id: UUID4
    acronym: str
    legacy_id: str | None
    molecule: Molecule
    name: str
    taxid: int
    plan: MonopartitePlan | MultipartitePlan


class CreateOTU(Event):
    """An event that creates a new OTU."""

    data: CreateOTUData
    query: OTUQuery


class ExcludeAccessionData(EventData):
    """The data for a :class:`ExcludeAccession` event."""

    accession: str


class ExcludeAccession(Event):
    """An accession exclusion event.

    This event is emitted when a Genbank accession is not going to be allowed in the
    reference.
    """

    data: ExcludeAccessionData
    query: OTUQuery


class CreatePlanData(EventData):
    """The data for a :class:`CreatePlan` event."""

    plan: MonopartitePlan | MultipartitePlan


class CreatePlan(Event):
    """An event that sets the isolate plan for an OTU."""

    data: CreatePlanData
    query: OTUQuery


class SetReprIsolateData(EventData):
    """The data for a :class:`SetReprIsolate` event."""

    isolate_id: UUID4


class SetReprIsolate(Event):
    """An event that sets the representative isolate for an OTU."""

    data: SetReprIsolateData
    query: OTUQuery
