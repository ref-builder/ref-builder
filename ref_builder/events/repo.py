from pydantic import UUID4

from ref_builder.events.base import EventData, Event, RepoQuery
from ref_builder.resources import RepoSettings


class CreateRepoData(EventData):
    """The data for a :class:`CreateRepo` event."""

    id: UUID4
    data_type: str
    name: str
    organism: str
    settings: RepoSettings


class CreateRepo(Event):
    """An event that creates a new repository.

    This event is always the first event in a repository's event log.
    """

    data: CreateRepoData
    query: RepoQuery
