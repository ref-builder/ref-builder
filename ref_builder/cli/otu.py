import sys
from pathlib import Path
from uuid import UUID

import click
from click import Context
import structlog

from ref_builder.cli.validate import validate_no_duplicate_accessions
from ref_builder.console import print_otu, print_otu_list
from ref_builder.options import ignore_cache_option, path_option
from ref_builder.otu.create import create_otu
from ref_builder.otu.isolate import (
    add_unnamed_isolate,
    add_and_name_isolate,
    add_genbank_isolate,
)
from ref_builder.otu.modify import (
    exclude_accessions_from_otu,
    set_representative_isolate,
    add_segments_to_plan,
    resize_monopartite_plan,
)
from ref_builder.otu.update import auto_update_otu, promote_otu_accessions
from ref_builder.otu.utils import RefSeqConflictError
from ref_builder.plan import SegmentRule, SegmentName
from ref_builder.repo import Repo
from ref_builder.utils import IsolateNameType, IsolateName

logger = structlog.get_logger()


@click.group(name="otu")
def otu() -> None:
    """Manage OTUs."""


@otu.command(name="create")
@click.argument("TAXID", type=int)
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option("--acronym", type=str, default="")
@ignore_cache_option
@path_option
def otu_create(
    accessions_: list[str],
    acronym: str,
    ignore_cache: bool,
    path: Path,
    taxid: int,
) -> None:
    """Create a new OTU.

    OTUs are created from a list of accessions and a taxonomy ID. The associated Genbank
    records are fetched and used to create the first isolate and a plan.
    """
    repo = Repo(path)

    if len(accessions_) != len(set(accessions_)):
        click.echo("Duplicate accessions were provided.", err=True)
        sys.exit(1)

    try:
        create_otu(
            repo,
            taxid,
            accessions_,
            acronym=acronym,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@otu.command(name="get")
@click.argument("IDENTIFIER", type=str)
@path_option
def otu_get(identifier: str, path: Path) -> None:
    """Get an OTU by its unique ID or taxonomy ID."""
    try:
        identifier = int(identifier)
        otu_ = Repo(path).get_otu_by_taxid(identifier)
    except ValueError:
        otu_ = Repo(path).get_otu(UUID(identifier))

    if otu_ is None:
        click.echo("OTU not found.", err=True)
        sys.exit(1)

    print_otu(otu_)


@otu.command(name="list")
@path_option
def otu_list(path: Path) -> None:
    """List all OTUs in the repository."""
    print_otu_list(Repo(path).iter_minimal_otus())


@otu.group(invoke_without_command=True)
@click.argument("TAXID", type=int)
@path_option
@click.pass_context
def update(ctx: Context, path: Path, taxid: int) -> None:
    """Update the specified OTU with new data."""
    repo = Repo(path)

    ctx.ensure_object(dict)
    ctx.obj = {
        "REPO": repo,
        "TAXID": taxid,
    }

    if not repo.get_otu_id_by_taxid(taxid):
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if ctx.invoked_subcommand is None:
        click.echo(ctx.get_help())


@update.command(name="automatic")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_auto_update(ctx: Context, ignore_cache: bool) -> None:
    """Automatically update an OTU with the latest data from NCBI."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        click.echo(f'Run "virtool otu create {taxid} --autofill" instead.')
        sys.exit(1)

    auto_update_otu(repo, otu_, ignore_cache=ignore_cache)


@update.command(name="promote")  # type: ignore
@ignore_cache_option
@click.pass_context
def otu_promote_accessions(
    ctx: Context,
    ignore_cache: bool,
) -> None:
    """Promote all RefSeq accessions within this OTU."""
    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(ctx.obj["TAXID"])

    promote_otu_accessions(repo, otu_, ignore_cache)


@update.command(name="isolate")  # type: ignore
@click.argument(
    "accessions_",
    callback=validate_no_duplicate_accessions,
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--unnamed",
    is_flag=True,
    default=False,
    help="ignore isolate names in Genbank sources",
)
@click.option(
    "--name",
    type=(IsolateNameType, str),
    help='an overriding name for the isolate, e.g. "isolate ARWV1"',
)
@ignore_cache_option
@click.pass_context
def isolate_create(
    ctx: Context,
    accessions_: list[str],
    ignore_cache: bool,
    name: tuple[IsolateNameType, str] | None,
    unnamed: bool,
) -> None:
    """Create a new isolate using the given accessions."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    if unnamed:
        add_unnamed_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )

    if name is not None:
        isolate_name_type, isolate_name_value = name
        isolate_name_ = IsolateName(type=isolate_name_type, value=isolate_name_value)

        try:
            add_and_name_isolate(
                repo,
                otu_,
                accessions_,
                ignore_cache=ignore_cache,
                isolate_name=isolate_name_,
            )
        except RefSeqConflictError as err:
            click.echo(
                f"{err.isolate_name} already exists, but RefSeq items may be "
                "promotable.",
            )
            sys.exit(1)

    try:
        add_genbank_isolate(
            repo,
            otu_,
            accessions_,
            ignore_cache=ignore_cache,
        )
    except RefSeqConflictError as err:
        click.echo(
            f"{err.isolate_name} already exists, but RefSeq items may be promotable,",
        )
        sys.exit(1)

    except ValueError as e:
        click.echo(e, err=True)
        sys.exit(1)


@update.command(name="exclude")  # type: ignore
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.pass_context
def accession_exclude(
    ctx: Context,
    accessions_: list[str],
) -> None:
    """Exclude the given accessions from this OTU.

    Any duplicate accessions will be ignored. Only one exclusion per unique accession
    will be made.
    """
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    exclude_accessions_from_otu(repo, otu_, accessions_)


@update.command(name="default")  # type: ignore
@click.argument("ISOLATE_KEY", type=str)
@click.pass_context
def otu_set_representative_isolate(
    ctx: Context,
    isolate_key: str,
) -> None:
    """Update the OTU with a new representative isolate."""
    taxid = ctx.obj["TAXID"]

    repo = ctx.obj["REPO"]

    otu_ = repo.get_otu_by_taxid(taxid)

    if otu_ is None:
        click.echo(f"OTU not found for Taxonomy ID {taxid}.", err=True)
        sys.exit(1)

    try:
        isolate_id = UUID(isolate_key)
    except ValueError:
        parts = isolate_key.split(" ")
        try:
            isolate_name = IsolateName(IsolateNameType(parts[0].lower()), parts[1])
        except ValueError:
            click.echo(f'Error: "{isolate_key}" is not a valid isolate name.', err=True)
            sys.exit(1)

        isolate_id = otu_.get_isolate_id_by_name(isolate_name)

    if isolate_id is None:
        click.echo("Isolate could not be found in this OTU.", err=True)
        sys.exit(1)

    set_representative_isolate(repo, otu_, isolate_id)


@update.group()
@click.pass_context
def plan(ctx: Context) -> None:
    """Add to and replace isolate plans for this OTU."""


@plan.command(name="extend")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_extend_segment_list(
    ctx: Context,
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Add recommended or optional segments to the OTU plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        add_segments_to_plan(
            repo,
            otu_,
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)


@plan.command(name="multipartite")
@click.argument(
    "accessions_",
    metavar="ACCESSIONS",
    nargs=-1,
    type=str,
    required=True,
)
@click.option(
    "--name",
    type=(str, str),
    help='a name for the originalsegment, e.g. "RNA A"',
)
@click.option(
    "--optional",
    is_flag=True,
    help="Set additional segments as fully optional",
)
@click.pass_context
@ignore_cache_option
def plan_resize_monopartite(
    ctx: Context,
    name: tuple[str, str],
    accessions_: list[str],
    optional: bool,
    ignore_cache: bool,
) -> None:
    """Change a monopartite OTU plan to a multipartite plan."""
    repo = ctx.obj["REPO"]
    taxid = ctx.obj["TAXID"]

    prefix, key = name

    otu_ = repo.get_otu_by_taxid(taxid)
    if otu_ is None:
        click.echo(f"OTU {taxid} not found.", err=True)
        sys.exit(1)

    try:
        resize_monopartite_plan(
            repo,
            otu_,
            name=SegmentName(prefix=prefix, key=key),
            rule=SegmentRule.OPTIONAL if optional else SegmentRule.RECOMMENDED,
            accessions=accessions_,
            ignore_cache=ignore_cache,
        )
    except ValueError as e:
        click.echo(e, err=True)
