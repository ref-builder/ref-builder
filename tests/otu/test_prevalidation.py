import pytest
from syrupy import SnapshotAssertion
from syrupy.filters import props

from ref_builder.ncbi.models import NCBISourceMolType
from ref_builder.otu.utils import assign_records_to_segments
from ref_builder.repo import Repo
from tests.fixtures.factories import NCBIGenbankFactory, NCBISourceFactory


@pytest.mark.parametrize("taxid", [223262, 3158377])
class TestAssignRecordsToSegments:
    """Test whether a list of records can be assigned to plan segments."""

    def test_ok(
        self,
        ncbi_genbank_factory: NCBIGenbankFactory,
        ncbi_source_factory: NCBISourceFactory,
        scratch_repo: Repo,
        snapshot: SnapshotAssertion,
        taxid: int,
    ):
        """Test that a list of records generated to match the OTU plan
        can pass assign_records_to_segments().
        """
        otu = scratch_repo.get_otu_by_taxid(taxid)

        records = [
            ncbi_genbank_factory.build(
                source=ncbi_source_factory.build(
                    mol_type=NCBISourceMolType.from_molecule(otu.molecule),
                    segment=str(segment.name),
                    organism=otu.name,
                    taxid=otu.taxid,
                )
            )
            for segment in otu.plan.required_segments
        ]

        assigned_records = assign_records_to_segments(records, otu.plan)

        assert sorted(assigned_records.values(), key=lambda r: r.accession) == snapshot(
            exclude=props("id")
        )

    def test_names_not_in_plan(
        self,
        ncbi_genbank_factory: NCBIGenbankFactory,
        scratch_repo: Repo,
        snapshot: SnapshotAssertion,
        taxid: int,
    ):
        """Test that a randomly generated list of records raises an appropriate
        ValueError.
        """
        otu = scratch_repo.get_otu_by_taxid(taxid)

        records = ncbi_genbank_factory.batch(len(otu.plan.required_segments))

        with pytest.raises(ValueError, match="Segment names not found in plan:") as e:
            assign_records_to_segments(records, otu.plan)

        assert str(e.value) == snapshot(exclude=props("id"))
