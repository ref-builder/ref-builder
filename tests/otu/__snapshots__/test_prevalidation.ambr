# serializer version: 1
# name: TestAssignRecordsToSegments.test_fail[223262]
  'Segment names not found in plan: DNA G, RNA N5.'
# ---
# name: TestAssignRecordsToSegments.test_fail[3158377]
  'Segment names not found in plan: DNA G, RNA A, RNA N5, Segment I3, Segment W4, Segment X.'
# ---
# name: TestAssignRecordsToSegments.test_ok[223262]
  dict({
    UUID('016e795b-f106-49cf-bdc5-fab1a978ccd9'): NCBIGenbank(accession='NC_907439', accession_version='NC_907439.2', strandedness=<Strandedness.SINGLE: 'single'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.LINEAR: 'linear'>, definition='XUbbCWtlvblwzmmlxpbR', organism='East African cassava mosaic Cameroon virus', sequence='KACGTGGHCDCYATHCSTGCGTCTGGGACAAGTGCTGCCCATAAKCTTGTYCCGTGBSCAGAWCTCSTGTACGTCGCCAACGMASAGCGTTGATATWCCTGCATTGTTGHDTHTTABTCTGHCGCCAGGCTGAAAAAYTACACGCARCGGGCAAACGCCSTCACTTAGGCATTATCYGCGCAAAACTTCTTATSGGCCGTTTNAAACKACCWTTCAGAATCATGGAGCACGAGTDGCGCGTATAGKGTTTATCCTGGGGCTAAATGAGACTTDAAGCKTACGAAGTTGGAAGTCGAGTCCAAAGTGGCAACGGGCCCSVATCAGTGGTTTTCAGACGMHAGCCAVTCAACATGKMCTTGBGTCGCDAKGCTCGATTCAGCAACCTATSGGCTCTGCNCATAGAHDCDGAGTGSRTYTCCATCYCTTCBCCAHTTAAATGGTGGATTAAGAGRACATCGCATGTGGTAGTCTGCAKTCNCGGGAKCCSGVGTCACGCDTCHGDAVTDTAGAGTTAAGAYGTGTCTACTACAATAGAACTTCTTAGYCCCATGNCTGTTCTCGTCTCGGTMMTTAGGCCMAWCAGGTTTAAGATTMBAGAGGCCGGTCTCAATCGHTTCTCTCAAAAACGGTAAGCGATGCCTACTACCAMGTACATGATTTTCTCGATGCDRCCGCKATCDAAAAATACYCCATAACGGGTGTTTTGATGTTAGAGCAKCTGCAGCCGCCAHNGATTTAATGCTTTGCAATGGCCTCTGTGACAACGTTCTVGTTCTCATAAAGSCGHGGAACCGGAGCTCRCGAACGTCTYACAGHTACTGACCCTTVABGCCGCNTGCNTTSMCCCCATAGHTATCATTCGAAAATTGCGGTGCAANGTAGWTSTTTTYAGWTTAAAGAAGACAATBAATACKGGGGNGAGGAACGATGCTCCACGATGMCGCGYGCGAGGTGGACTRACTTATTCRAGCWTCTCCCTAGGCCACCAGVACCAGGTCGAAACCCTTGGGAGTCGTGTATAGACTANTTCTAKTGGATAGBAKGGAMTACGCTMTATACATGTACSTAGGAHTCGCCCTAACGGCGGTTAGGTHATGCWCAGAGSAATGTTTCAGCGGGCCGCGCDTCGMCAWGAAGTGTCGYCTTGTGTCATACCCGGACCACWTTTHTTGGTCCSTTACTAATGCSCTCGKACATCGGGTGNCTTAACACCATAATACACCNATCCAAAGACGAGCG', source=NCBISource(taxid=223262, organism='East African cassava mosaic Cameroon virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='4_8217', host='squash', segment='DNA A', strain='', clone='', proviral=False, macronuclear=True, focus=False, transgenic=False), comment='igDQOHZPCKKQzoWnojLe', refseq=True),
    UUID('a1f7b2a7-45d9-4ff8-a131-b3e0ad718251'): NCBIGenbank(accession='K93863', accession_version='K93863.2', strandedness=<Strandedness.DOUBLE: 'double'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.LINEAR: 'linear'>, definition='hxKeWhIbvvhFICWShkLu', organism='East African cassava mosaic Cameroon virus', sequence='TGTCAGACTGCCGTAGTGBCAGCTACARGGTKAGTAAAGTCCTCCKCGDGTTAGACCAAGCKRAGATTBCACCATGHGARCGAARTACVGTNGGCTRCMCYCTCTACGTGGGTNGCACARTTCCCCCGTCGAATAGAATCGNGCAACGTDCCCAASTATGTATCGTGHGATKGATTGATGGACCGTHAAACATAABTDGAMAVATCAGAGACATCTTCAAGTMTCAAACCWTGTATGGWGCCCCGCGCCGGCCGCTTCACGTTGKGCTTCTGAGGTATCGGAAAVCAGBCTBGAGCCTTTCCGAASCCCTTAAVGGAGTCGATCGCTCCTCTATACCGGGVTTTACCAKHAAAGGAACTAAAATCTTAGTCCTTAACGAARAGGAGAATTCTGCAAGAWGATTTTATATCGCCGTTTGGYCAAGGCKMGTCTTCAAACGGACCAACTCAGTTCGYGGAACVGATB', source=NCBISource(taxid=223262, organism='East African cassava mosaic Cameroon virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='6Eritrea', host='grapevine', segment='DNA B', strain='', clone='', proviral=False, macronuclear=False, focus=False, transgenic=False), comment='ATSOGIrgXboFYkmSrusc', refseq=False),
  })
# ---
# name: TestAssignRecordsToSegments.test_ok[3158377]
  dict({
    UUID('5a3a379e-533d-42df-b1ce-099bc9066e4c'): NCBIGenbank(accession='NC_934447', accession_version='NC_934447.2', strandedness=<Strandedness.DOUBLE: 'double'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.CIRCULAR: 'circular'>, definition='GOsuddzxpnRRJzwTwNWL', organism='abaca bunchy top virus', sequence='GGTABATTTTGTTANGCAACGTTGACCBAATATSATGAMCAAATGARKATGCCCGTAARCGATAGGYVTTCCGYGAACGTTAGWTCAGATCAGCTCCCCAAACCGCCTGTCAGCTNCTCAATAGCMTGACCTTNCAGGCTHYTGCANTAAGTCACACATCCCTCGGCCGRTDCAATGCCAGGASYTCGCGGCGGTTCTBATGTGTDCGYAABGCTGTCRTACGGAGTGGAAGGTCCAGHCGSNHGACMTCCNTCGTACGAGCNTGGAGSTTGTGGTGATTGAAACCHCATNCTCGTGAGKVAMGTTCTGACSACTAGCACAACTGCTCCATCCGTAWTWATCAGTNACAACCYMTAGTTTATGGTCCRTCAAATGCTTGASCAACGATTVATHTGCGCGTMGGTGCCCGAAAAACAATYTCGCCGTGATNTGGTACCTGCTCTCGGGCGGATAAGAGGTCGGGAGGCGCTCACVCTGGGTCCCTGTTTATYGTNGT', source=NCBISource(taxid=3158377, organism='abaca bunchy top virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='Johnson_GN', host='squash', segment='DNA M', strain='tell-rest', clone='', proviral=False, macronuclear=False, focus=False, transgenic=False), comment='VqinUmRJuheyLPVpuaZg', refseq=True),
    UUID('7079086b-b32d-48d1-a47f-267bc5c347bb'): NCBIGenbank(accession='NC_907439', accession_version='NC_907439.2', strandedness=<Strandedness.SINGLE: 'single'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.LINEAR: 'linear'>, definition='XUbbCWtlvblwzmmlxpbR', organism='abaca bunchy top virus', sequence='KACGTGGHCDCYATHCSTGCGTCTGGGACAAGTGCTGCCCATAAKCTTGTYCCGTGBSCAGAWCTCSTGTACGTCGCCAACGMASAGCGTTGATATWCCTGCATTGTTGHDTHTTABTCTGHCGCCAGGCTGAAAAAYTACACGCARCGGGCAAACGCCSTCACTTAGGCATTATCYGCGCAAAACTTCTTATSGGCCGTTTNAAACKACCWTTCAGAATCATGGAGCACGAGTDGCGCGTATAGKGTTTATCCTGGGGCTAAATGAGACTTDAAGCKTACGAAGTTGGAAGTCGAGTCCAAAGTGGCAACGGGCCCSVATCAGTGGTTTTCAGACGMHAGCCAVTCAACATGKMCTTGBGTCGCDAKGCTCGATTCAGCAACCTATSGGCTCTGCNCATAGAHDCDGAGTGSRTYTCCATCYCTTCBCCAHTTAAATGGTGGATTAAGAGRACATCGCATGTGGTAGTCTGCAKTCNCGGGAKCCSGVGTCACGCDTCHGDAVTDTAGAGTTAAGAYGTGTCTACTACAATAGAACTTCTTAGYCCCATGNCTGTTCTCGTCTCGGTMMTTAGGCCMAWCAGGTTTAAGATTMBAGAGGCCGGTCTCAATCGHTTCTCTCAAAAACGGTAAGCGATGCCTACTACCAMGTACATGATTTTCTCGATGCDRCCGCKATCDAAAAATACYCCATAACGGGTGTTTTGATGTTAGAGCAKCTGCAGCCGCCAHNGATTTAATGCTTTGCAATGGCCTCTGTGACAACGTTCTVGTTCTCATAAAGSCGHGGAACCGGAGCTCRCGAACGTCTYACAGHTACTGACCCTTVABGCCGCNTGCNTTSMCCCCATAGHTATCATTCGAAAATTGCGGTGCAANGTAGWTSTTTTYAGWTTAAAGAAGACAATBAATACKGGGGNGAGGAACGATGCTCCACGATGMCGCGYGCGAGGTGGACTRACTTATTCRAGCWTCTCCCTAGGCCACCAGVACCAGGTCGAAACCCTTGGGAGTCGTGTATAGACTANTTCTAKTGGATAGBAKGGAMTACGCTMTATACATGTACSTAGGAHTCGCCCTAACGGCGGTTAGGTHATGCWCAGAGSAATGTTTCAGCGGGCCGCGCDTCGMCAWGAAGTGTCGYCTTGTGTCATACCCGGACCACWTTTHTTGGTCCSTTACTAATGCSCTCGKACATCGGGTGNCTTAACACCATAATACACCNATCCAAAGACGAGCG', source=NCBISource(taxid=3158377, organism='abaca bunchy top virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='4_8217', host='squash', segment='DNA N', strain='', clone='', proviral=False, macronuclear=True, focus=False, transgenic=False), comment='igDQOHZPCKKQzoWnojLe', refseq=True),
    UUID('773df54d-275b-4dcb-9432-7b9c1f5b6ffe'): NCBIGenbank(accession='NC_404705', accession_version='NC_404705.2', strandedness=<Strandedness.SINGLE: 'single'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.LINEAR: 'linear'>, definition='oGnMioACeKggqRypNnCF', organism='abaca bunchy top virus', sequence='TGGACGTACATCACGCCAACTCACCGAGAKGGKATCTCGATGTAGTCACGTTGCCAGGATCAGATCARCTCGCCGTTACTWTGTATCCTTACGRCTGCGAGTCAAAAAGCANTGCTGDTTTGCAACAGCAACAHCTGCCGCAANCAA', source=NCBISource(taxid=3158377, organism='abaca bunchy top virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='Paraguay_Turner', host='tomato', segment='DNA C', strain='', clone='', proviral=False, macronuclear=False, focus=False, transgenic=False), comment='fnIWDzVgUHmNjExfWktH', refseq=True),
    UUID('986487e2-fc3d-4cc0-84a3-18b65e46563d'): NCBIGenbank(accession='NC_761067', accession_version='NC_761067.2', strandedness=<Strandedness.DOUBLE: 'double'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.LINEAR: 'linear'>, definition='YapytkYCWAOHhKMLzcRY', organism='abaca bunchy top virus', sequence='GATTTCCGCTAVGAMGAGTAAAGASGGAVTGHTAGTTACAAGACTCGGATTAATGTTRAGCGCTTGDTMMATTCGCATGGCCCAATTGCTCCATGTGCTACAACGGCGTGGTACGTGCTTCAATATTGAHCHATCACCGCCCAADACGCAGRAGACGTTTMTCTHCTGGTGTBCACCCTGGBAVTACCTGCTAACTGCTGTCAAGAGGACGGCSGGATCTTTAWTGTGTTTGTGAT', source=NCBISource(taxid=3158377, organism='abaca bunchy top virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='HallKG', host='garlic', segment='DNA R', strain='', clone='', proviral=False, macronuclear=False, focus=False, transgenic=False), comment='ZEFMaXruxCtzfUVUZWym', refseq=True),
    UUID('d9ba6203-c8c3-4d43-a23b-1c1232e2888d'): NCBIGenbank(accession='K93863', accession_version='K93863.2', strandedness=<Strandedness.DOUBLE: 'double'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.LINEAR: 'linear'>, definition='hxKeWhIbvvhFICWShkLu', organism='abaca bunchy top virus', sequence='TGTCAGACTGCCGTAGTGBCAGCTACARGGTKAGTAAAGTCCTCCKCGDGTTAGACCAAGCKRAGATTBCACCATGHGARCGAARTACVGTNGGCTRCMCYCTCTACGTGGGTNGCACARTTCCCCCGTCGAATAGAATCGNGCAACGTDCCCAASTATGTATCGTGHGATKGATTGATGGACCGTHAAACATAABTDGAMAVATCAGAGACATCTTCAAGTMTCAAACCWTGTATGGWGCCCCGCGCCGGCCGCTTCACGTTGKGCTTCTGAGGTATCGGAAAVCAGBCTBGAGCCTTTCCGAASCCCTTAAVGGAGTCGATCGCTCCTCTATACCGGGVTTTACCAKHAAAGGAACTAAAATCTTAGTCCTTAACGAARAGGAGAATTCTGCAAGAWGATTTTATATCGCCGTTTGGYCAAGGCKMGTCTTCAAACGGACCAACTCAGTTCGYGGAACVGATB', source=NCBISource(taxid=3158377, organism='abaca bunchy top virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='6Eritrea', host='grapevine', segment='DNA U3', strain='', clone='', proviral=False, macronuclear=False, focus=False, transgenic=False), comment='ATSOGIrgXboFYkmSrusc', refseq=False),
    UUID('dd46de05-6f66-4d43-b40d-82857b741e91'): NCBIGenbank(accession='GG223149', accession_version='GG223149.1', strandedness=<Strandedness.DOUBLE: 'double'>, moltype=<MolType.DNA: 'DNA'>, topology=<Topology.CIRCULAR: 'circular'>, definition='UCUCwTEyAswdkNgkiscs', organism='abaca bunchy top virus', sequence='TGATCCAGGGMAGCRGCTCCCMCTTGATGTNCTVYTTGGWTAATATCTGGCTTAAATGGAVACGACAACCCTCAGTTGGCCTGTTTCCGTGAAMATTGCCCGCTAATCTAAAGGCTGCCTCCTGACGGTTGCAGTCTAAACATCCCCWCCTTGGTGAACHCTTTGTGTTTATGGCGCTTAAGVGCAGTCVCSTTGTGCTGTAAAAATNCGMTCCGTGGRACGMATCATATCGGARNRTNTGACDTGCGTGTAGAATGWAGGATTTGSCTSAACDATCTGCGGCAGTGCACACNAGTGVTACATGGGTTCAAGGGTTMCGCTTCMTTDKAATTTGTAAGCAGTATGTTTGCAKTAAGCAWGTAACCGHCACCAGGTCWTNTGTAACGATGAGCATDACATTCCAYCAYGGWTACGTTCTHTDGVCTGAANCGATGTVGACACACCTTGAATGGTCGCATATCGCGACTGAACGACGACTACGGTCTGACANTDGTGAAGMHGWTSCAATGAATGAAGGTCGMCCTGTHAACCGTCATCATCATTCACGSCCTDGATGTATTACCTCGTSCCTCGCGATTAAAGTCGGCTGGCTACTACTAGAGCTACTTGGTCTTTGABTCGCCTCTDGATAGDGGATGATTAGGGTVCTAGACGCCATAATCTTKTAGCGTATCACAACACGHTGCGACYAAGTCCTTCTVTTCCAGCACCCGATATGCCGGACCCCATTTCMTCGGTAACTTAAWGTBSACCAGTAACCCTGTWTTTGTRGAGCTAGCHCACWTTATAGCCTCTTTTGACATCTCATACTGDGGCGACACATTACGTCTGTGKATAAAAATCGCATGTCBAAGTTTTCACGTGACGGCTTCBSCTTTCGGCTGAGTTGCCACAAAACGCATCGCTTTTTCTTGCVTRGCAGTTTCGTRAGYCCCGTCCAAAGCTTGCATATCTCTCTCAAGGWAAAACGCGGGCGGCAGACCTATCMACTACKGTSGCGGGATACGAMCTDAGASTGTCGCTGCGAGCCYTCCKCDAGACCTCACTGSTADDGKGCATTTAMGCATATAGTGCACCVSTCGGCACACGBGTCACGCCCGTTGCTCACGAGTNGAATGTCCSAGCCTTRTTCGTHTGTTACCTCCGAATATCWAATGTTGATCGCTYGAACTTTVGGAATCGTNGTTCGCMKTCGGGACACTGTTTATGAYCTAYCAGCCDGGTAGCCTCAACGTTATGHTGTCGGGTAAMAGAGBCTKTCCSTATCAGGTCGGTCACGCRATVATTCCCCTCGASCGWGGTTTAAGATTTAGCGCACCCTGTTCAAGCTCAKGCCGGGAGGCAACCT', source=NCBISource(taxid=3158377, organism='abaca bunchy top virus', mol_type=<NCBISourceMolType.GENOMIC_DNA: 'genomic DNA'>, isolate='', host='apple', segment='DNA S', strain='', clone='', proviral=False, macronuclear=False, focus=False, transgenic=False), comment='InAnuIGULLnEoHOjIfoM', refseq=False),
  })
# ---
